<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coronavirus Monitoring for Inclusivity Tool • CoMIT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Coronavirus Monitoring for Inclusivity Tool">
<meta property="og:description" content="CoMIT builds a sqlite3 database from primer and sequence data.
    Primer inputs are submitted via Excel file and then parsed to create a 
    primer table which is used to analyze sequences for mismatches compared 
    to the primers.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CoMIT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="comit">CoMIT<a class="anchor" aria-label="anchor" href="#comit"></a>
</h1></div>
<p>CoMIT is an R package developed by the Applied Math team at BioFire Defense. The goal of CoMIT is to monitor performance of BioFire Defense’s COVID-19 7-assay test with reference to SARS-CoV-2 sequences and metadata pulled from the GISAID EpiCoV database (<a href="https://gisaid.org/" class="external-link uri">https://gisaid.org/</a>). It uses a binning algorithm to sort sequences based on mutation characteristics and performs pairwise alignments to the reference sequence only when absolutely necessary. Output is organised in a SQLite database. A subset database can be pulled from the original database, taking a user-created lineage csv file as input. This subset database is then used to create various visualizations, with sequences categorized based on lineage from the input file. Sequences are also grouped in the visualizations according to an internal system for evaluating predicted risk to the COVID-19 test.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level4">
<h4 id="from-source">From Source<a class="anchor" aria-label="anchor" href="#from-source"></a>
</h4>
<p>You can install CoMIT like so, where “path” is the path to the source (.tar.gz) file:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"CoMIT_1.0.0.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
<p>If a message appears saying that dependencies are not available, then install those dependencies by running <code>install.packages("dependency name")</code> in the console. Then re-run the installation of the source file.</p>
<p>If you get the following warning:</p>
<p><code>WARNING: Rtools is required to build R packages but is not currently installed. Please download and install the appropriate version of Rtools before proceeding.</code></p>
<p>then your computer doesn’t have Rtools installed, which is required for installing source packages. First run <code>R.version</code> in the console and take note of your version of R (it should show up next to the version.string variable). Then go <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">here</a> and click on the Rtools version that corresponds to your R version. You should be fine to install Rtools to the default location. Once the installer finishes, you should be able to install CoMIT.</p>
</div>
<div class="section level4">
<h4 id="from-github">From GitHub<a class="anchor" aria-label="anchor" href="#from-github"></a>
</h4>
<p>To install CoMIT from GitHub, run the following line of code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"BFDFAppliedMath/CoMIT"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>Here are the exported functions from the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Main algorithm functions</span></span>
<span><span class="fu"><a href="reference/build_COVID_DB.html">build_COVID_DB</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/update_db.html">update_db</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/pull_subset_db.html">pull_subset_db</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization functions</span></span>
<span><span class="fu"><a href="reference/makeAssaySummaryTables.html">makeAssaySummaryTables</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makeAssaySummaryTablesCondensed.html">makeAssaySummaryTablesCondensed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makeCoOccHeatmaps.html">makeCoOccHeatmaps</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makeCoOccTables.html">makeCoOccTables</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makeCoOccTablesCondensed.html">makeCoOccTablesCondensed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makePrimerVarTable.html">makePrimerVarTable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makedbBreakdownTable.html">makedbBreakdownTable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/makedbBreakdownTableCondensed.html">makedbBreakdownTableCondensed</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Useful utility functions</span></span>
<span><span class="fu"><a href="reference/buildDBFromSchema.html">buildDBFromSchema</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/mismatchGenerator.html">mismatchGenerator</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/translateMutationForward.html">translateMutationForward</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/translateMutationReverse.html">translateMutationReverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>To see all the functions available for use, run <code>getNamespaceExports(“CoMIT”)</code>. To see information on a specific function, including a description and its required parameters, run <code>?functionName</code>. Note that all visualization functions begin with the word “make”.</p>
<p>Here is the general pipeline for running CoMIT:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a new database: </span></span>
<span><span class="fu"><a href="reference/build_COVID_DB.html">build_COVID_DB</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Update a database with sequences from one or more .fasta file by running them against the BioFire defense Covid-19 test primers (run the sorting algorithm): </span></span>
<span><span class="fu"><a href="reference/update_db.html">update_db</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a subset database from the original database with a lineage file:</span></span>
<span><span class="fu"><a href="reference/pull_subset_db.html">pull_subset_db</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run whatever visualization functions you would like to:</span></span>
<span><span class="co"># (any exported function beginning with "make")</span></span></code></pre></div>
<p>And here is an example of building a new database:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bfdfappliedmath.github.io/CoMIT/" class="external-link">CoMIT</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/build_COVID_DB.html">build_COVID_DB</a></span><span class="op">(</span>DB_NAME <span class="op">=</span> <span class="st">"test"</span>, locationPrefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span>, include_all_assays <span class="op">=</span> <span class="cn">TRUE</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, silently <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "The following tables were built:"</span></span>
<span><span class="co">#&gt;  [1] "Ambiguous_Seqs"    "Assay_Info"        "Cospot_Key"       </span></span>
<span><span class="co">#&gt;  [4] "High_N_Seqs"       "ID_Map"            "Primer_Details"   </span></span>
<span><span class="co">#&gt;  [7] "Primer_Info"       "Primer_Locations"  "Seq_Info"         </span></span>
<span><span class="co">#&gt; [10] "Seq_Meta"          "Var_Amplicon_Info" "Variant_Info"     </span></span>
<span><span class="co">#&gt; [13] "Variant_Type"      "sqlite_sequence"</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpzgUn3t/test.db"</span></span></code></pre></div>
<p>Here are some examples of visualizations that can be produced. These come from the BioFire Defense monthly evaluation which uses sequences collected over the span of three months and submitted to the GISAID EpiCoV database.</p>
<p><img src="reference/figures/assay_summary_hr.png"></p>
<p><img src="reference/figures/co_occ_hr.png"></p>
<p><img src="reference/figures/heatmap_hr.png"></p>
<p><img src="reference/figures/database_breakdown.png"></p>
<p><img src="reference/figures/primer_variant_vis.png"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://creativecommons.org/licenses/by/4.0" class="external-link">CC BY 4.0</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing CoMIT</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ashlie Stephenson <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Ashlie Stephenson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
