<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ CoMIT</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CoMIT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/CoMIT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addVariant2DB.html">addVariant2DB()</a></code> </p>
        </td>
        <td><p>Insert new Variant into a CoMIT database and add to cache$primerVars</p></td>
      </tr><tr><td>
          <p><code><a href="add_extra_metadata.html">add_extra_metadata()</a></code> </p>
        </td>
        <td><p>Insert location and sequencing information</p></td>
      </tr><tr><td>
          <p><code><a href="buildDBFromSchema.html">buildDBFromSchema()</a></code> </p>
        </td>
        <td><p>Build a database from a given SQL schema file</p></td>
      </tr><tr><td>
          <p><code><a href="build_COVID_DB.html">build_COVID_DB()</a></code> </p>
        </td>
        <td><p>Build COVID-19 Analysis Database</p></td>
      </tr><tr><td>
          <p><code><a href="classifyFromAlignment.html">classifyFromAlignment()</a></code> </p>
        </td>
        <td><p>Classify Variant from a masked alignment</p></td>
      </tr><tr><td>
          <p><code><a href="classifyVariant.html">classifyVariant()</a></code> </p>
        </td>
        <td><p>Classify an identified variant sequence and insert into Variant Info and Variant Type tables</p></td>
      </tr><tr><td>
          <p><code><a href="comit_classify.html">comit_classify()</a></code> </p>
        </td>
        <td><p>COMIT Variant Classifier</p></td>
      </tr><tr><td>
          <p><code><a href="formatDF.html">formatDF()</a></code> </p>
        </td>
        <td><p>Format Database Breakdown Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="formatDF_2.html">formatDF_2()</a></code> </p>
        </td>
        <td><p>Format Database Breakdown Dataframe V2</p></td>
      </tr><tr><td>
          <p><code><a href="format_as_gt_tbl.html">format_as_gt_tbl()</a></code> </p>
        </td>
        <td><p>Format Assay Summary GT Table</p></td>
      </tr><tr><td>
          <p><code><a href="format_as_gt_tbl_2.html">format_as_gt_tbl_2()</a></code> </p>
        </td>
        <td><p>Format Assay Summary GT Table V2</p></td>
      </tr><tr><td>
          <p><code><a href="format_co_gt_tbl.html">format_co_gt_tbl()</a></code> </p>
        </td>
        <td><p>Format Co-occurring Mutations GT Table</p></td>
      </tr><tr><td>
          <p><code><a href="format_co_gt_tbl_2.html">format_co_gt_tbl_2()</a></code> </p>
        </td>
        <td><p>Format Co-occurring Mutations GT Table V2</p></td>
      </tr><tr><td>
          <p><code><a href="getAssaySummaryCounts.html">getAssaySummaryCounts()</a></code> </p>
        </td>
        <td><p>Get Assay Summary Counts</p></td>
      </tr><tr><td>
          <p><code><a href="getAssaySummaryCounts_2.html">getAssaySummaryCounts_2()</a></code> </p>
        </td>
        <td><p>Get Assay Summary Counts V2</p></td>
      </tr><tr><td>
          <p><code><a href="getCoOccCounts.html">getCoOccCounts()</a></code> </p>
        </td>
        <td><p>Get Co-Occurring Counts</p></td>
      </tr><tr><td>
          <p><code><a href="getCoOccCounts5to7.html">getCoOccCounts5to7()</a></code> </p>
        </td>
        <td><p>Get Co-occurring Counts for 5 to 7 Assays</p></td>
      </tr><tr><td>
          <p><code><a href="getCoOccCounts5to7_2.html">getCoOccCounts5to7_2()</a></code> </p>
        </td>
        <td><p>Get Co-occurring Counts for 5 to 7 Assays V2</p></td>
      </tr><tr><td>
          <p><code><a href="getCoOccCounts_2.html">getCoOccCounts_2()</a></code> </p>
        </td>
        <td><p>Get Co-Occurring Counts V2</p></td>
      </tr><tr><td>
          <p><code><a href="getMaskedAlignmentList.html">getMaskedAlignmentList()</a></code> </p>
        </td>
        <td><p>Get Masked Alignment List from supplied primers</p></td>
      </tr><tr><td>
          <p><code><a href="getMetaData.html">getMetaData()</a></code> </p>
        </td>
        <td><p>Parse metadata from a Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="getThreeAssayProportion.html">getThreeAssayProportion()</a></code> </p>
        </td>
        <td><p>Get Three-Assay Proportion</p></td>
      </tr><tr><td>
          <p><code><a href="group_block_anno.html">group_block_anno()</a></code> </p>
        </td>
        <td><p>Add Group Block Annotation</p></td>
      </tr><tr><td>
          <p><code><a href="makeAssaySummaryTables.html">makeAssaySummaryTables()</a></code> </p>
        </td>
        <td><p>Make Assay Summary Tables</p></td>
      </tr><tr><td>
          <p><code><a href="makeAssaySummaryTablesCondensed.html">makeAssaySummaryTablesCondensed()</a></code> </p>
        </td>
        <td><p>Make Assay Summary Tables (Condensed Version)</p></td>
      </tr><tr><td>
          <p><code><a href="makeCoOccHeatmaps.html">makeCoOccHeatmaps()</a></code> </p>
        </td>
        <td><p>Make Heatmaps for Co-occurring mutations</p></td>
      </tr><tr><td>
          <p><code><a href="makeCoOccTables.html">makeCoOccTables()</a></code> </p>
        </td>
        <td><p>Make Tables for Co-occurring mutations</p></td>
      </tr><tr><td>
          <p><code><a href="makeCoOccTablesCondensed.html">makeCoOccTablesCondensed()</a></code> </p>
        </td>
        <td><p>Make Tables for Co-occurring mutations (Condensed Version)</p></td>
      </tr><tr><td>
          <p><code><a href="makeCombinedDataframe.html">makeCombinedDataframe()</a></code> </p>
        </td>
        <td><p>Make Combined Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="makeCombinedDataframe_2.html">makeCombinedDataframe_2()</a></code> </p>
        </td>
        <td><p>Make Combined Dataframe V2</p></td>
      </tr><tr><td>
          <p><code><a href="makeDFcoOcc.html">makeDFcoOcc()</a></code> </p>
        </td>
        <td><p>Make Co-occurring Mutations Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="makeDFcoOccHeatmap.html">makeDFcoOccHeatmap()</a></code> </p>
        </td>
        <td><p>Make Dataframe for Co-occurring Mutations Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="makeDFcoOcc_2.html">makeDFcoOcc_2()</a></code> </p>
        </td>
        <td><p>Make Co-occurring Mutations Dataframe V2</p></td>
      </tr><tr><td>
          <p><code><a href="makeDFdatabaseBreakdown.html">makeDFdatabaseBreakdown()</a></code> </p>
        </td>
        <td><p>Make Database Breakdown Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="makeDFdatabaseBreakdown_2.html">makeDFdatabaseBreakdown_2()</a></code> </p>
        </td>
        <td><p>Make Database Breakdown Dataframe V2</p></td>
      </tr><tr><td>
          <p><code><a href="makePrimerVarFreqLinDFs.html">makePrimerVarFreqLinDFs()</a></code> </p>
        </td>
        <td><p>Make primer variant frequency and lineage dataframes</p></td>
      </tr><tr><td>
          <p><code><a href="makePrimerVarMutationDF.html">makePrimerVarMutationDF()</a></code> </p>
        </td>
        <td><p>Make Primer Variant Mutation Dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="makePrimerVarTable.html">makePrimerVarTable()</a></code> </p>
        </td>
        <td><p>Make primer variant visualization</p></td>
      </tr><tr><td>
          <p><code><a href="makedbBreakdownTable.html">makedbBreakdownTable()</a></code> </p>
        </td>
        <td><p>Make Database Breakdown Table</p></td>
      </tr><tr><td>
          <p><code><a href="makedbBreakdownTableCondensed.html">makedbBreakdownTableCondensed()</a></code> </p>
        </td>
        <td><p>Make Database Breakdown Table (Condensed Version)</p></td>
      </tr><tr><td>
          <p><code><a href="mismatchGenerator.html">mismatchGenerator()</a></code> </p>
        </td>
        <td><p>Generate Mismatches for a primer sequence</p></td>
      </tr><tr><td>
          <p><code><a href="parseFastaLine.html">parseFastaLine()</a></code> </p>
        </td>
        <td><p>Parse GISAID FASTA line for metadata and check for sequences with no accession number,
the wrong host, and/or invalid collection dates</p></td>
      </tr><tr><td>
          <p><code><a href="pull_subset_db.html">pull_subset_db()</a></code> </p>
        </td>
        <td><p>Pull subset DB</p></td>
      </tr><tr><td>
          <p><code><a href="reorderVect.html">reorderVect()</a></code> </p>
        </td>
        <td><p>Reorder vect - change the position of what in vect to where</p></td>
      </tr><tr><td>
          <p><code><a href="setSubsetStrings.html">setSubsetStrings()</a></code> </p>
        </td>
        <td><p>Set strings in "subset" environment for pull_subset_DB function</p></td>
      </tr><tr><td>
          <p><code><a href="translateMutationForward.html">translateMutationForward()</a></code> </p>
        </td>
        <td><p>Translate a Seq Change String to a Primer Template Mismatch for Forward primers</p></td>
      </tr><tr><td>
          <p><code><a href="translateMutationReverse.html">translateMutationReverse()</a></code> </p>
        </td>
        <td><p>Translate a Seq Change String to a Primer Template Mismatch for Reverse primers</p></td>
      </tr><tr><td>
          <p><code><a href="update_db.html">update_db()</a></code> </p>
        </td>
        <td><p>Update DB (Add sequence info to a database)</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ashlie Stephenson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

